app/code/Magento/MediaGalleryUi/Model/Directories/GetDirectoryTree.php

```diff
@@ -77,25 +85,46 @@ private function getDirectories(): array
         /** @var Read $directory */
         $directory = $this->filesystem->getDirectoryRead(DirectoryList::MEDIA);
 
-        if (!$directory->isDirectory()) {
-            return $directories;
-        }
-
-        foreach ($directory->readRecursively() as $path) {
-            if (!$directory->isDirectory($path) || $this->isPathExcluded->execute($path)) {
-                continue;
+        $isExcluded = function (string $path) use ($directory): bool {
+            return $this->isPathExcluded->execute($directory->getRelativePath($path));
+        };
+        $flags = FilesystemIterator::SKIP_DOTS |
+            FilesystemIterator::UNIX_PATHS |
+            RecursiveDirectoryIterator::FOLLOW_SYMLINKS;
+
+        try {
+            $iterator = new RecursiveIteratorIterator(
+                new RecursiveCallbackFilterIterator(
+                    new RecursiveDirectoryIterator($directory->getAbsolutePath(), $flags),
+                    static function (
+                        SplFileInfo $file,
+                        $key,
+                        RecursiveIterator $iterator
+                    ) use ($isExcluded): bool {
+                        return $file->isDir() && $iterator->hasChildren() && !$isExcluded($file->getPathname());
+                    }
+                ),
+                RecursiveIteratorIterator::CHILD_FIRST
+            );
+            /** @var FilesystemIterator $file */
+            foreach ($iterator as $file) {
+                $path = $directory->getRelativePath($file->getPathname());
+                $directories[] = [
+                    'data' => $file->getBasename(),
+                    'attr' => ['id' => $file->getPathname()],
+                    'metadata' => [
+                        'path' => $path
+                    ],
+                    'path_array' => explode('/', $path)
+                ];
```

Comment:
If the sample is validated I can implement a new FileDriver and override the readDirectoryRecursively to allow RecursiveCallbackFilterIterator

Reply:
@thomas-kl1 Hi, i thinks that any performance fixes for this issue is like just temporary fixes, so there is only one way to fix this issue, is to update behavior to load folders on a first level path, and then render sub folders by ajax requests and get one by one as it was in old media gallery, and this is available for the jstree plugin, instead of reading the whole directory tree. If you have enough time to contribute you can implement this behavior as well, what's do you think ?

Reply:
@Nazar65 that's true, anyway I've noticed that `RecursiveIteratorIterator` was used in Magento framework. However this should be used without a filter. I'm looking to add the possibility to give the filter as an argument to the method `readRecursively`.
Regarding the ajax, I agree it's the best solution to implement. I may have time this week to work on it.

Reply:
@thomas-kl1 cool :+1:

Reply:
@Nazar65 I didn't had the time yet as it requires much more attention and development. I see how it was done in the old media gallery and I need to be sure it will be backward-compatible

Reply:
@thomas-kl1  you don't need to change anything just provide a new component and then just change the jstree initialization, it's not a BC changes providing new functionality